### ISSUE TITLE
On background save, Redis responds very very slowly...

### ISSUE THREAD

**Issue by GoogleCodeExporter** [ashish21590/redis#154]
```
What steps will reproduce the problem?
1. Store 275,000 keys
2. Store keys at rate of 30,000 per minute
3. As soon as Redis goes to background save, connections take 6+ seconds to 
respond

What version of the product are you using? On what operating system?
1.2.0 on Centos 5.4 x86

Please provide any additional information below.

Here is from the redis log:
06 Feb 06:12:39 - 200000 changes in 300 seconds. Saving...
06 Feb 06:12:39 - Background saving started by pid 14154
06 Feb 06:13:03 - DB saved on disk

Here is from the PHP log
string(76) "2/6/10 6:12:49 Redis didn't not respond to socket_select"
string(76) "2/6/10 6:12:52 Redis didn't not respond to socket_select"
string(76) "2/6/10 6:12:56 Redis didn't not respond to socket_select"

The timeout on socket_select is 6 seconds. Could the problem be that I have 
compression turned on? Would it help to turn it off?

I have also increased the auto save count greatly in the mean time.

Thanks,
James Hartig

```

Original issue reported on code.google.com by `fastest...@gmail.com` on 6 Feb 2010 at 5:20

**Comment by GoogleCodeExporter** [comment]
```
Hello, it sounds simpler than that, I think your system is swapping while the 
system
is saving because you are nearly out of memory. Can you check this with the "ps"
utility? Not only while saving, you can call ps while Redis is working as 
usually to
check if as I suspect it's using almost all the available RAM.

Cheers,
Salvatore
```

Original comment by `anti...@gmail.com` on 6 Feb 2010 at 5:25

**Comment by GoogleCodeExporter** [comment]
```
We have 1.4GB free of memory right now, I don't think that is the issue. 
Current load 
is .02 .06 .06 so I don't think it is the CPU either.

Will I have to sit and keep monitoring for when it saves? It saves every 10-15 
minutes 
during the save you want me to call ps?
```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 5:34

**Comment by GoogleCodeExporter** [comment]
```
Also, I have upgraded to 1.2.1 to see if that helps.
```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 5:35

**Comment by GoogleCodeExporter** [comment]
```
Ok, but can you please report the amount of memory reported by 'ps'? This will 
be 
helpful, as it must be something at the OS level, because the saving child is a 
completely different process, and there is no known way it can interact with 
the main 
process. Don't trust the memory reported by INFO, it can be much larger due to 
malloc fragmentation.

Even if you are actually not out of memory, maybe you have a big part of the 
dataset 
rarely used and memory pages about it are swapped on disk, and loaded back when 
you do a BGSAVE.

So to track this issue the information needed are: complete INFO output, memory 
reported by ps.

Also I've some trouble parsing this numbers:

1. Store 275,000 keys
2. Store keys at rate of 30,000 per minute

How you can have 270,000 keys if you store 30,000 keys per minute? Keys have 
EXPIREs set? You keep setting the same keys again and again?

Thanks for any info,
Salvatore
```

Original comment by `anti...@gmail.com` on 6 Feb 2010 at 5:40

**Comment by GoogleCodeExporter** [comment]
```
Sorry typo,
We are currently at 275,000+ keys
We are making 30,000 COMMANDs per minute. The DB is reset at 4am EST time so 
the 
275,000 is low because well it was just reset a little bit ago.

Also, I do not have the privilege of setting vm.overcommit_memory.

INFO
$379
redis_version:1.2.1
arch_bits:32
multiplexing_api:epoll
uptime_in_seconds:1085
uptime_in_days:0
connected_clients:5
connected_slaves:0
used_memory:53094852
changes_since_last_save:162096
bgsave_in_progress:0
last_save_time:1265478194
bgrewriteaof_in_progress:0
total_connections_received:13
total_commands_processed:2024914
role:master
db0:keys=278403,expires=38

From what this says, we have a little more than 30k per minute, however on 
average it 
is 30k but every few hours it goes through and does a mass-transfer which gets 
up to 
300k per minute and lasts for a minute or two.

Current memory usage: 
RSS: 74148kb
VSZ: 78028kb

During save: (I'm still looking at redis-server right?)
RSS: 80288kb
VSZ: 78772kb

What is odd is that it saved really quickly this time:
06 Feb 12:58:15 - 500000 changes in 900 seconds. Saving...
06 Feb 12:58:15 - Background saving started by pid 19936
06 Feb 12:58:16 - DB saved on disk
06 Feb 12:58:17 - Background saving terminated with success

I understand that it is a separate process, I was not trying to blame you or 
Redis, 
I'm just trying to figure out what is causing this. The times are always 
perfect with 
the saves that redis does and so it has something to do with saving.

Thanks

```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 6:00

**Comment by GoogleCodeExporter** [comment]
```
Sorry forgot to include the most recent INFO (after save)

INFO
$379
redis_version:1.2.1
arch_bits:32
multiplexing_api:epoll
uptime_in_seconds:1962
uptime_in_days:0
connected_clients:5
connected_slaves:0
used_memory:56168744
changes_since_last_save:145327
bgsave_in_progress:0
last_save_time:1265479097
bgrewriteaof_in_progress:0
total_connections_received:15
total_commands_processed:3892177
role:master
db0:keys=281229,expires=59

```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 6:01

**Comment by GoogleCodeExporter** [comment]
```
Hello again,

no problem I just need data to guess what's happening. Indeed your Redis 
instance is 
using very little memory, this behavior is really strange. I wonder if when 
this happens 
you can monitor what is going on with vmstat and iostat Linux commands.

Btw never saw this behavior before with a dataset perfectly fitting RAM.
```

Original comment by `anti...@gmail.com` on 6 Feb 2010 at 6:22

**Comment by GoogleCodeExporter** [comment]
```
Alright I will look at vmstat and iostat.

Could you look into a timeout for the saving client? If the save lasts longer 
than 5 
seconds cancel the save. For most people this is not what they want, but we 
have 100% 
uptime and if I lose the last 15 mins of data, I personally have more important 
things 
to care about.
```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 7:35

**Comment by GoogleCodeExporter** [comment]
```
We do not have any swap partition:
cat /proc/meminfo
MemTotal:      2097152 kB
MemFree:       1728136 kB
Buffers:             0 kB
Cached:              0 kB
SwapCached:          0 kB
Active:              0 kB
Inactive:            0 kB
HighTotal:           0 kB
HighFree:            0 kB
LowTotal:      2097152 kB
LowFree:       1728136 kB
SwapTotal:           0 kB
SwapFree:            0 kB
Dirty:               0 kB

Current RAM readings:
RSS: 108108kb
VSZ: 109608kb

INFO
$380
redis_version:1.2.1
arch_bits:32
multiplexing_api:epoll
uptime_in_seconds:8243
uptime_in_days:0
connected_clients:5
connected_slaves:0
used_memory:76057625
changes_since_last_save:97958
bgsave_in_progress:0
last_save_time:1265485419
bgrewriteaof_in_progress:0
total_connections_received:47
total_commands_processed:16526001
role:master
db0:keys=297850,expires=100

Since the saves are <2 seconds so far, I have not been able to accurately get 
the 
vmstat and iostat outputs in the time that I realize the save started. So far I 
have 
not got a single failed command from PHP though.
```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 7:47

**Comment by GoogleCodeExporter** [comment]
```
Ok, but I've an idea that could tell at least if it's really a matter of the 
background 
saving or not. What about if you run this script?

while [ 1 ]; do echo "$(date) $(redis-cli info | grep uptime_in_seconds)"; 
sleep 1; done

The output will be like:

Sab  6 Feb 2010 21:13:11 CET uptime_in_seconds:110
Sab  6 Feb 2010 21:13:12 CET uptime_in_seconds:111
Sab  6 Feb 2010 21:13:13 CET uptime_in_seconds:112
Sab  6 Feb 2010 21:13:14 CET uptime_in_seconds:113

Redirect it to some file, so you can be sure Redis was not responding to 
clients. 
Because if I understand correctly there is no evidence that it's not a bug in 
the 
application.

If it's a problem with Redis really not responding we'll see that there is an 
"hole" in 
the log output produced by INFO. If we can't see any hole, we know there is 
something to look about the application.
```

Original comment by `anti...@gmail.com` on 6 Feb 2010 at 8:15

**Comment by GoogleCodeExporter** [comment]
```
Is there anyway to run redis-cli with a password? 

Yes you are correct it could be a bug in the application.
```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 8:24

**Comment by GoogleCodeExporter** [comment]
```
Unfortunately not :(

This is something I should add in the TODO list.

But you can send any command and remove the grep, even if will reply with error 
no 
problem, we just need any kind of reply from it, to see if it's alive.
```

Original comment by `anti...@gmail.com` on 6 Feb 2010 at 8:29

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Alright it is running. Once I notice a PHP error I will report what was in the 
redis 
log and in the new uptime log at that same time. So far today though (after 
increases 
in save time and I doubled the PHP timeouts) there has not been any errors.
```

Original comment by `fastest...@gmail.com` on 6 Feb 2010 at 8:39

**Comment by GoogleCodeExporter** [comment]
```
Appears to be fixed in 1.2.2
```

Original comment by `fastest...@gmail.com` on 21 Feb 2010 at 9:22

**Comment by GoogleCodeExporter** [comment]
```
@fastest963: thanks for reporting, after the fix and your tweet about upgrading 
1.2.2 I 
was wondering if this was what you were observing indeed. And sounds like it 
was 
indeed.
```

Original comment by `anti...@gmail.com` on 21 Feb 2010 at 9:35
- Changed state: **Verified**
